subprojects/core/src/integTest/groovy/org/gradle/api/file/FileContentsIntegrationTest.groovy

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.file
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Issue
+
+import java.nio.charset.Charset
+
+@Issue("https://github.com/gradle/gradle/issues/15802")
+class FileContentsIntegrationTest extends AbstractIntegrationSpec  {
```

Comment:
:question: Maybe add a test for the invalid encoding name too? And, while we're there, resolve encoding eagerly, so the request for the invalid encoding fails when `getAsText` is called rather than when the provider is materialized.

We're not very consistent about eagerness of checking the encoding names, but I believe the eager check provides a better UX.